superset-frontend/src/explore/controlUtils/standardizedFormData.ts

```diff
@@ -0,0 +1,182 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import {
+  ensureIsArray,
+  getChartControlPanelRegistry,
+  QueryFormData,
+} from '@superset-ui/core';
+import {
+  ControlStateMapping,
+  SharedFormData,
+  iStandardizedFormData,
+} from '@superset-ui/chart-controls';
+import { getControlsState } from 'src/explore/store';
+import { getFormDataFromControls } from './getFormDataFromControls';
+
+export const sharedControls: Record<keyof SharedFormData, string[]> = {
+  metrics: ['metric', 'metrics', 'metric_2'],
+  columns: ['groupby', 'columns', 'groupbyColumns', 'groupbyRows'],
+};
+export const publicControls = [
+  // time section
+  'granularity_sqla',
+  'time_grain_sqla',
+  'time_range',
+  // filters
+  'adhoc_filters',
+  // subquery limit(series limit)
+  'limit',
+  // order by clause
+  'timeseries_limit_metric',
+  'series_limit_metric',
+  // desc or asc in order by clause
+  'order_desc',
+  // outer query limit
+  'row_limit',
+  // x asxs column
+  'x_axis',
+  // advanced analytics - rolling window
+  'rolling_type',
+  'rolling_periods',
+  'min_periods',
+  // advanced analytics - time comparison
+  'time_compare',
+  'comparison_type',
+  // advanced analytics - resample
+  'resample_rule',
+  'resample_method',
+];
+
+export class StandardizedFormData {
+  private sfd: iStandardizedFormData;
+
+  constructor(sourceFormData: QueryFormData) {
+    /*
+     * Support form_data for smooth switching between different viz
+     * */
+    const sharedFormData = {
+      metrics: [],
+      columns: [],
+    };
+    const formData = { ...sourceFormData };
+    const reversedMap = StandardizedFormData.getReversedMap();
+
+    Object.entries(formData).forEach(([key, value]) => {
+      if (reversedMap.has(key)) {
+        sharedFormData[reversedMap.get(key)].push(...ensureIsArray(value));
+      }
+    });
+
+    const memorizedFormData = Array.isArray(
+      formData?.standardized_form_data?.memorizedFormData,
+    )
+      ? new Map(formData.standardized_form_data.memorizedFormData)
+      : new Map();
+    const vizType = formData.viz_type;
+    if (memorizedFormData.has(vizType)) {
+      memorizedFormData.delete(vizType);
+    }
+    memorizedFormData.set(vizType, formData);
+    this.sfd = {
+      sharedFormData,
+      memorizedFormData,
+    };
+  }
+
+  static getReversedMap() {
+    const reversedMap = new Map();
+    Object.entries(sharedControls).forEach(([key, names]) => {
+      names.forEach(name => {
+        reversedMap.set(name, key);
+      });
+    });
+    return reversedMap;
+  }
+
+  private getLatestFormData(vizType: string): QueryFormData {
+    if (this.sfd.memorizedFormData.has(vizType)) {
+      return this.sfd.memorizedFormData.get(vizType) as QueryFormData;
+    }
+
+    return this.memorizedFormData.slice(-1)[0][1];
```

Comment:
Should we add some null check in case `memorizedFormData` is empty?

Reply:
the `memorizedFormData` will not be empty, it must be initialized in the constructor.